### ISSUE TITLE
A different way of doing the similarity/comparison task?

### ISSUE THREAD

**Issue by chaitjo** [IndicoDataSolutions/finetune#131]
Hey! Thanks for the awsome work. I was wondering if I could use and update finetune to do the following:

Instead of using `(Start, Text1, Delim, Text2, Extract)` and `(Start, Text2, Delim, Text1, Extract)` as in the paper, can we use `(Start, Text1, Extract)` and `(Start, Text2, Extract)` separately through the transformer? 

This could be thought of as obtaining sentence/document embeddings for Text1 and Text2 separately. Upon doing that, I would like to compare their similarity using a distance metric such as cosine distance. (i.e. train the transformer as a siamese network.)

Would you suggest I build such a model on top of a fork of finetune?

**Comment by madisonmay** [comment]
Hi @chaitjo!

Thanks for opening an issue -- your suggestion is a reasonable one.  The current implementation is so only because the implementation follows the [diagram on page 4 in the original paper](https://s3-us-west-2.amazonaws.com/openai-assets/research-covers/language-unsupervised/language_understanding_paper.pdf).

You should be able to achieve what you're looking for by inheriting from the `Comparison` class and overriding [the _text_to_ids method](https://github.com/IndicoDataSolutions/finetune/blob/development/finetune/comparison.py#L20. ) and the corresponding [_target_model method](https://github.com/IndicoDataSolutions/finetune/blob/development/finetune/classifier.py#L65).

If you demonstrate that the modification is a useful one (i.e. outperforms the current implementation on a few tasks) I would be all for adding an additional model type to `finetune` :)

**Comment by chaitjo** [comment]
Thanks for the response! Will keep you posted on the progress.

**Comment by madisonmay** [comment]
Sounds good!  If you get stuck and think a walkthrough of the relevant code would be helpful happy to setup a screenshare or something. Happy hacking!

**Comment by chaitjo** [comment]
Hey @madisonmay! I've worked on something based on your initial pointers:
- Added a `cosine_similarity` target model to the `network_modules` which takes as input two hidden states and produces the cosine similarity (Dot product -> sigmoid) between them.
- This meant that I have to add a new `siamese_comparison` class which inherits from the `BaseModel`, because I needed to take two sets of `Xs` as input and update the training loop as well as the graph construction for this. Most methods had to be overwritten.

[Here's the relevant branch](https://github.com/chaitjo/finetune/tree/cosine_similarity_model).

Am I using the codebase correctly? I have yet to make this run, but what data should I test this out on?

**Comment by madisonmay** [comment]
Hey @chaitjo, glad you've been hacking away over the weekend!  [I've opened up a PR](https://github.com/IndicoDataSolutions/finetune/pull/142/files) into development for you -- don't worry if things don't run yet I'll just use it as a space to leave code comments for now.  

At a high level things look good, seems like there are just a few things to clean up.  Perhaps most importantly, I think we can find a way around overriding many of the `BaseModel` methods by structuring things more like the comparison class (and maybe inheriting from that instead of the `BaseModel` class.)

As far as data goes, a sampled version of the [Quora similarity dataset](https://github.com/IndicoDataSolutions/finetune/blob/development/finetune/datasets/quora_similarity.py) might be a good place to start.   We've got some scripts in there for training already, but note that you'll probably have to modify those scripts to convert the pandas Series passed as inputs to numpy arrays -- that's on my backlog to patch up.  E.g.:

```
    dataset = QuoraDuplicate(nrows=100).dataframe
    model = Comparison(verbose=True, n_epochs=3, lm_warmup=0.1)
    trainX1, testX1, trainX2, testX2, trainY, testY = train_test_split(dataset.Text1, dataset.Text2, dataset.Target, test_size=0.3, random_state=42)
    model.fit(list(zip(trainX1.values, trainX2.values)), trainY.values)
    accuracy = np.mean(model.predict(list(zip(testX1.values, testX2.values))) == testY.values)
    class_balance = np.mean(testY.values)
    print('Test Accuracy: {:0.2f} for a {:0.2f} class balance'.format(accuracy, class_balance))
```

**Comment by madisonmay** [comment]
How did your research go @chaitjo?

**Comment by madisonmay** [comment]
Closing due to lack of activity.

**Comment by chaitjo** [comment]
Hey, sorry for the late response. The library worked well and I was able to customize it to my needs easily: various siamese network styles + cosine similarity. Research ongoing.

**Comment by nstfk** [comment]
@chaitjo Hi, I also wanted to implement a similar siamese model, and thank to @benleetownsend for pointing this out!
I can see that your fork is no longer available, but on a similar task did you achieve better results with the siamese network vs the mutiFieldClassifier ?
it would be great to share your thoughts !

**Comment by chaitjo** [comment]
@nstfk I was able to get something to work in a hack-y way, but I did not evaluate it on the QQP dataset. Nor did I compare it to the multiFieldClassifier. 

Sorry for the late reply and for not being very helpful here. I think you could use the helpful cues by the maintainers to go about implementing this.

**Comment by madisonmay** [comment]
@nstfk we haven't benchmarked with a contrastive loss / Siamese net, but we're currently working on a GLUE submission over at https://github.com/IndicoDataSolutions/finetune/tree/ben/add_explicit_val_for_glue and have gotten solid results using the `Comparison` model in finetune (~0.90 acc.).
